sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala

```
@@ -77,106 +61,76 @@ class AccountForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
-    }
-    // 2. Compute the receipt root
-    val receiptsRoot: Array[Byte] = calculateReceiptRoot(receiptList)
-
-    // 3. As forger address take first address from the wallet
+      nodeView: View,
+      branchPointInfo: BranchPointInfo,
+      nextConsensusEpochNumber: ConsensusEpochNumber,
+      isWithdrawalEpochLastBlock: Boolean,
+      parentId: BlockId,
+      timestamp: Timestamp,
+      mainchainBlockReferencesData: Seq[MainchainBlockReferenceData],
+      sidechainTransactions: Seq[SidechainTypes#SCAT],
+      mainchainHeaders: Seq[MainchainHeader],
+      ommers: Seq[Ommer[AccountBlockHeader]],
+      ownerPrivateKey: PrivateKey25519,
+      forgingStakeInfo: ForgingStakeInfo,
+      vrfProof: VrfProof,
+      forgingStakeInfoMerklePath: MerklePath,
+      companion: DynamicTypedSerializer[SidechainTypes#SCAT, TransactionSerializer[SidechainTypes#SCAT]],
+      inputBlockSize: Int,
+      signatureOption: Option[Signature25519]
+  ): Try[SidechainBlockBase[SidechainTypes#SCAT, AccountBlockHeader]] = {
+
+    // As forger address take first address from the wallet
     val addressList = nodeView.vault.secretsOfType(classOf[PrivateKeySecp256k1])
     if (addressList.size() == 0)
       throw new IllegalArgumentException("No addresses in wallet!")
-
     val forgerAddress = addressList.get(0).publicImage().asInstanceOf[AddressProposition]
 
-    // TODO: 4. calculate baseFee
+    // TODO: calculate baseFee
     val baseFee = 0
+    val gasLimit = Account.GAS_LIMIT
 
-    // 5. Get cumulativeGasUsed from last receipt in list if available
-    val gasUsed = if (receiptList.size > 0) receiptList.last.cumulativeGasUsed.longValue() else 0
+    val blockContext = new BlockContext(
+      forgerAddress.address(),
+      timestamp,
+      baseFee,
+      gasLimit,
+      nodeView.history.getBlockHeightById(parentId).asScala.map(height => height + 1).getOrElse(0),
```

I would say that in case there is no parent, out block has height `1`, not `0`.
So can be `nodeView.history.getBlockHeightById(parentId).asScala.getOrElse(0) + 1`

---

And will the genesis block ever go through this function?
- if no, there will be no block that does not have a parent so it doesnt matter..?
- if yes, what is the height of the genesis block, `0` or `1`?

---

genesis block does not use it.
but the height of the genesis block is 1.
I even think, that in case Forger get `nodeView.history.getBlockHeightById(parentId)` result as None -> it is a bug an we need to throw an exception.